app/gui/src/project-view/util/codemirror/markdownEditing.ts

```diff
@@ -0,0 +1,285 @@
+
+/** Supported header levels. */
+export type HeaderLevel = 1 | 2 | 3
+
+/** Toggle headers of specified level at each of the selected lines. */
+export function toggleHeader(view: EditorView, level: HeaderLevel) {
+  const selection = view.state.selection.main
+  const startLine = view.state.doc.lineAt(selection.from)
+  const endLine = view.state.doc.lineAt(selection.to)
+  const src = view.state.doc.toString()
+  const tree = markdownParser.parse(src)
+  const changeSet = new MutableChangeSet(0)
+  for (let lineIndex = startLine.number; lineIndex <= endLine.number; lineIndex++) {
+    const line = view.state.doc.line(lineIndex)
+    const context = new Context(tree, src, line)
+    const lineChanges = toggleHeaderInner(context, level)
+    changeSet.merge(lineChanges)
+  }
+  changeSet.dispatch(view, true)
+}
+
+function toggleHeaderInner(context: Context, level: number): MutableChangeSet {
+  const prefix = `${'#'.repeat(level)} `
+  const node = resolveNodeAtPos(context.tree, context.lineEnd())
+  const changeSet = context.makeChangeSet()
+  if (isCodeText(node)) {
+    const codeText = context.src.slice(node.from, node.to)
+    const codeTree = markdownParser.parse(codeText)
+    const codeContext = new Context(codeTree, codeText, context.line).withOffset(node.from)
+    const codeChanges = toggleHeaderInner(codeContext, level)
+    changeSet.merge(codeChanges)
+  } else {
+    const headerMark = findAtxHeaderMark(node)
+    if (headerMark && headerMark.level === level) {
+      changeSet.remove(headerMark.from, headerMark.to)
+    } else if (headerMark) {
+      changeSet.replace(headerMark.from, headerMark.to, prefix)
+    } else {
+      changeSet.add(context.lineStart(), context.lineStart(), prefix)
+    }
+  }
+  return changeSet
+}
+
+function findAtxHeaderMark(node: SyntaxNode): { level: number; from: number; to: number } | null {
+  const cursor = node.cursor()
+  do {
+    if (cursor.type.name.startsWith('ATXHeading')) {
+      const headerMark = cursor.node.getChild('HeaderMark')
+      if (!headerMark) return null
+      const level = Number(cursor.type.name.slice(-1))
+      return { level, from: headerMark.from, to: headerMark.to }
+    }
+  } while (cursor.parent())
+  return null
+}
+
+// =============
+// === Lists ===
+// =============
+
+/** Distinguish between unordered (bullet) and ordered (numbered) lists. */
+export type ListType = 'unordered' | 'ordered'
+
+/** Toggle list items of specified type at each of the selected lines. */
+export function toggleList(view: EditorView, type: ListType) {
+  const tree = markdownParser.parse(view.state.doc.toString())
+  const startLine = view.state.doc.lineAt(view.state.selection.main.from)
+  const endLine = view.state.doc.lineAt(view.state.selection.main.to)
+  const changeSet = new MutableChangeSet(0)
+  let listIndex = 0
+  for (let i = startLine.number; i <= endLine.number; i++) {
+    const line = view.state.doc.line(i)
+    const src = view.state.doc.toString()
```

Comment:
This builds a string from the entire document text, do we need to do it in a loop?

Reply:
Nice catch, it was a leftover from refactoring. Fixed.